src/debug/debug_stream/debug_stream_thread_info.c

```diff
@@ -0,0 +1,371 @@
+ * time and it moves to the cpu its pinned on.
+ */
+static void cleanup_old_thread_cycles(struct user_data *ud)
+{
+	int i, j;
+
+	for (i = 0; i < ARRAY_SIZE(ud->previous->threads); i++) {
+		bool found = false;
+
+		/* This entry is already free, continue */
+		if (ud->previous->threads[i].tid == NULL)
+			continue;
+
+		/* Check if the thread was seen on previous round */
+		for (j = 0; j < ud->thread_count; j++) {
+			if (ud->active_threads[j] == ud->previous->threads[i].tid) {
+				found = true;
+				break;
+			}
+		}
+		/* If the thead is not any more active, mark the entry free */
+		if (!found) {
+			ud->previous->threads[i].tid = NULL;
+			ud->previous->threads[i].cycles = 0;
+		}
+	}
+}
+#else
+static void cleanup_old_thread_cycles(struct user_data *ud) { }
+#endif
+
+static void thread_info_get(int core, struct record_buf *rec)
+{
+	k_thread_runtime_stats_t core_stats;
+	struct user_data ud = {
+		.core = core,
+		.rec = rec,
+		.thread_count = 0,
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+		.active_threads = { NULL },
+		.previous = &previous[core],
+#endif
+	};
+	struct thread_info_record_hdr *hdr;
+	uint8_t load = 0;
+#ifdef CONFIG_THREAD_RUNTIME_STATS
+	int ret = k_thread_runtime_stats_cpu_get(core, &core_stats);
+
+	if (ret == 0) {
+		uint32_t active_cycles = (uint32_t) (core_stats.total_cycles -
+						     ud.previous->active);
+		uint32_t all_cycles = (uint32_t) (core_stats.execution_cycles -
+						  ud.previous->all);
+
+		ud.stats_valid = true;
+		load = (uint8_t) ((255LLU * active_cycles) / all_cycles);
+		LOG_DBG("Core %u load %u / %u total %llu / %llu", core,
+			active_cycles, all_cycles,
+			core_stats.total_cycles, core_stats.execution_cycles);
+		ud.previous->active = core_stats.total_cycles;
+		ud.previous->all = core_stats.execution_cycles;
+		ud.all_cycles = all_cycles;
+	}
+#endif
+
+	hdr = (struct thread_info_record_hdr *) ud.rec->buf;
+	rec->w_ptr = sizeof(*hdr);
+	hdr->hdr.id = DEBUG_STREAM_RECORD_ID_THREAD_INFO;
+	hdr->load = load;
+	/* This is best effort debug tool. Unlocked version should be fine. */
+	k_thread_foreach_unlocked_filter_by_cpu(core, thread_info_cb, &ud);
+
+	cleanup_old_thread_cycles(&ud);
+
+	hdr->thread_count = ud.thread_count;
+	hdr->hdr.size = SOF_DIV_ROUND_UP(rec->w_ptr, sizeof(hdr->hdr.data[0]));
+	debug_stream_slot_send_record(&hdr->hdr);
+}
+
+static void thread_info_run(void *cnum, void *a, void *b)
+{
+	int core = (int) cnum;
+	struct record_buf rec = {
+		.size = THREAD_INFO_INITIAL_RECORD_BUFFER_SIZE,
+		.w_ptr = 0,
+	};
+
+	rec.buf = rmalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, rec.size);
+	if (!rec.buf) {
+		LOG_ERR("malloc failed");
+		return;
+	}
+
+	for (;;) {
+		thread_info_get(core, &rec);
+		k_sleep(K_SECONDS(CONFIG_SOF_DEBUG_STREAM_THREAD_INFO_INTERVAL));
+	}
+}
+
+#define THREAD_STACK_SIZE (2048)
+static K_THREAD_STACK_ARRAY_DEFINE(info_thread_stacks, CONFIG_MP_MAX_NUM_CPUS,
+				   THREAD_STACK_SIZE);
+static struct k_thread info_thread[CONFIG_MP_MAX_NUM_CPUS];
```

Comment:
`CONFIG_MP_NUM_CPUS`?

Reply:
Oh, that is obsolete. Will change to MP_MAX_NUM_CPUS Zephyr option, unles you now some better way to know number of CPUs we should prepare for at compile time. 

config MP_MAX_NUM_CPUS
        int "Maximum number of CPUs/cores"
        default 1
        range 1 12
        help
          Maximum number of multiprocessing-capable cores available to the
          multicpu API and SMP features.